src/function/scalar/list/list_select.cpp

```diff
@@ -0,0 +1,164 @@
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/function/scalar/nested_functions.hpp"
+#include "duckdb/planner/expression_binder.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/planner/expression/bound_parameter_expression.hpp"
+
+namespace duckdb {
+
+struct SetSelectionVectorSelect {
+	static void SetSelectionVector(SelectionVector &selection_vector, ValidityMask &validity_mask,
+	                               ValidityMask &input_validity, Vector &selection_entry, idx_t child_idx,
+	                               idx_t &target_offset, idx_t selection_offset, idx_t input_offset,
+	                               idx_t target_length) {
+		idx_t sel_idx = selection_entry.GetValue(selection_offset + child_idx).GetValue<int>() - 1;
+		if (sel_idx < target_length) {
+			selection_vector.set_index(target_offset, input_offset + sel_idx);
+			if (!input_validity.RowIsValid(input_offset + sel_idx)) {
+				validity_mask.SetInvalid(target_offset);
+			}
+		} else {
+			selection_vector.set_index(target_offset, 0);
+			validity_mask.SetInvalid(target_offset);
+		}
+		target_offset++;
+	}
+
+	static void GetResultLength(DataChunk &args, idx_t &result_length) {
+		idx_t count = args.size();
+		UnifiedVectorFormat unified_vec;
+		args.data[0].ToUnifiedFormat(count, unified_vec);
+		result_length = ListVector::GetListSize(args.data[1]) * unified_vec.validity.CountValid(count);
+	}
+};
+
+struct SetSelectionVectorWhere {
+	static void SetSelectionVector(SelectionVector &selection_vector, ValidityMask &validity_mask,
+	                               ValidityMask &input_validity, Vector &selection_entry, idx_t child_idx,
+	                               idx_t &target_offset, idx_t selection_offset, idx_t input_offset,
+	                               idx_t target_length) {
+		idx_t child_bool = selection_entry.GetValue(selection_offset + child_idx).GetValue<bool>();
+		if (child_bool) {
+			selection_vector.set_index(target_offset, input_offset + child_idx);
+			if (!input_validity.RowIsValid(input_offset + child_idx)) {
+				validity_mask.SetInvalid(target_offset);
+			}
+			target_offset++;
+		}
+	}
+
+	static void GetResultLength(DataChunk &args, idx_t &result_length) {
+		idx_t count = args.size();
+		UnifiedVectorFormat unified_vec;
+		args.data[0].ToUnifiedFormat(count, unified_vec);
+		result_length = ListVector::GetListSize(args.data[1]) * unified_vec.validity.CountValid(count);
+	}
+};
+template <class OP>
+static void ListSelectFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.data.size() == 2);
+	Vector &list = args.data[0];
+	Vector &selection_list = args.data[1];
+	idx_t count = args.size();
+	idx_t result_length;
+	OP::GetResultLength(args, result_length);
```

Comment:
Looking at the implementation of `GetResultLength`, I think they are incorrect for both `where` and `select`. 

`ListVector::GetListSize(args.data[1]) * unified_vec.validity.CountValid(count);`:
`ListVector::GetListSize(args.data[1])` returns the length of the child vector of the selection list, which is the length of the result child vector without accounting for `NULL` values in the input lists. BUT: `unified_vec.validity.CountValid(count);` is the count of non-null lists in the input list column. This means we multiply our selection list for every valid list, which is too much. Is that correct?

> I'm not quite happy with the result length calculation for `list_where'. I don't know how to calculate the real size of the result list without reading the list of bools twice. But maybe taking the length of the maximum possible list is ok.

I think that you've two options. 

1: You traverse the list and process vectors of `STANDARD_VECTOR_SIZE` (see `list_aggregates.cpp`). 

2: It also makes sense to traverse the input twice. The complexity class does not change, and this is a somewhat tight loop, so it should also be efficient. For `select`, we only count the select list's length towards the resulting length if the input list is not `NULL`. We count all true values towards the result length for `where` if the input list is not `NULL`. You can write a general loop for this and then use the functors for the specific code paths.

Reply:
I noticed something while implementing a good way to calculate the resulting length. If we have ConstantVector of NULL, should the ValidityMask be invalid everywhere? Because for one vector of the DataChunk it is not. Only the first entry in the VadilityMask is invalid, and all others are valid. (I could be completely wrong here :D)

Reply:
Our constant vectors have a selection vector that only ever retrieves the first entry (and its validity) for any index. We also propagate that to all possible (nested) child vectors. So, your observation makes sense! If you always use `.sel->get_index` to retrieve the actual data index in a vector and then use `.validity.RowIsValid` with that index, you should get the necessary behavior without seeing unexpected validity mask values. :)  

For example, you can see the `SetNull` function's functionality here and how we set the selection vector to always point to the 0-indexed element in `ZeroSelectionVector` (both in `vector.cpp`).
```cpp
//===--------------------------------------------------------------------===//
// ConstantVector
//===--------------------------------------------------------------------===//
void ConstantVector::SetNull(Vector &vector, bool is_null) {
	D_ASSERT(vector.GetVectorType() == VectorType::CONSTANT_VECTOR);
	vector.validity.Set(0, !is_null);
	if (is_null && vector.GetType().InternalType() == PhysicalType::STRUCT) {
		// set all child entries to null as well
		auto &entries = StructVector::GetEntries(vector);
		for (auto &entry : entries) {
			entry->SetVectorType(VectorType::CONSTANT_VECTOR);
			ConstantVector::SetNull(*entry, is_null);
		}
	}
}

const SelectionVector *ConstantVector::ZeroSelectionVector(idx_t count, SelectionVector &owned_sel) {
	if (count <= STANDARD_VECTOR_SIZE) {
		return ConstantVector::ZeroSelectionVector();
	}
	owned_sel.Initialize(count);
	for (idx_t i = 0; i < count; i++) {
		owned_sel.set_index(i, 0);
	}
	return &owned_sel;
}
```

Reply:
Makes sense!